Switch from getopt to argparse

evitalis opened the pull request:
I was able to run tox tests for linting and py{37,27}-ansible29 without issue.
This addresses the bulk of #635, the final warning cannot be resolved while Python 2.7 is supported.

webknjaz commented on lib/ansiblelint/__main__.py:
```diff
@@ -177,7 +180,11 @@ def main():
     options.skip_list = frozenset(skip)
 
-    playbooks = sorted(set(args))
+    if len(options.playbook) > 0:
```
Maybe just
```suggestion
playbooks = sorted(set(options.playbook or args))
```

webknjaz commented on lib/ansiblelint/__main__.py:
```diff
@@ -149,7 +152,7 @@ def main():
 
     # no args triggers auto-detection mode
-    if len(args) == 0 and not (options.listrules or options.listtags):
+    if len(options.playbook) == 0 and not (options.listrules or options.listtags):
```
```suggestion
    if not options.playbook and not (options.listrules or options.listtags):
```

evitalis commented:
@webknjaz Both changes made and committed. I have no idea why this now has conflicts. I added your suggestions and added a commit with those minor changes. I also tried to squash the commits which didn't resolve the conflict.

Checked my files with `git diff` before I committed too and didn't see anything that was wrong. Thoughts on what happened?

evitalis commented:
Fixed the conflict, 8b7a4c4b87d5d3856d017e44ea5cede5e20dc1cb ended up conflicting due to some edits in that commit. Should be fine now.

evitalis commented:
ping @webknjaz

webknjaz commented on lib/ansiblelint/__main__.py:
```diff
@@ -52,62 +52,64 @@ def main():
+                        help='Specify configuration file to use.  '
+                             'Defaults to ".ansible-lint"')
+    parser.add_argument('--version', action='version', version='%(prog)s ' + __version__)
+    parser.add_argument('playbook', nargs='*')
```
Why is this arg needed?

webknjaz commented on lib/ansiblelint/__main__.py:
```diff
@@ -52,62 +52,64 @@ def main():
+                             'Defaults to ".ansible-lint"')
+    parser.add_argument('--version', action='version', version='%(prog)s ' + __version__)
+    parser.add_argument('playbook', nargs='*')
+    options = parser.parse_args()
```
This guide https://docs.python.org/3.9/library/argparse.html#upgrading-optparse-code says that it should be
```suggestion
    args = parser.parse_args()
```

So why did you keep options? OTOH it probably contains both args and options in a single var instead of two vars.

webknjaz replied:
Also, why don't you keep `sys.argv[1:]`. Or even better — add an optional arg to `main()` to make custom in-script invocations possible.

evitalis commented:
@webknjaz I probably left both as they were there previously. That being said its been awhile so I no longer remember. Feel free to make what edits you feel are needed as I am not interested in contributing to this project further. As noted in #635 though the current code is deprecated and should be updated.

ssbarnea commented:
@evitalis Please rebase it, I cannot do it for you and this is likely the solution for closing #730

webknjaz commented:
@ssbarnea it'd be non-trivial to rebase this given that arg parsing stuff got moved to `cli` module by @cans recently.

ssbarnea commented:
@webknjaz feel free to ping me on irc if you want me to do a manual check of the change once done.

webknjaz commented:
I think this change is okay, I just need to adjust a small tests patch and will merge.

webknjaz commented:
This is tricky, postponing for now
